bot.py

```diff
@@ -457,13 +458,35 @@ async def countdown(self, message, command_content):
         except ValueError:
             await message.channel.send(f":robot: {command_content} is not an integer.")
 
+    async def interrupt_play(self, message, source):
+        """
+        Pause currently playing audio c (if any), play source s, then resume c
+        """
+        voice_client = await self.create_or_get_voice_client(message)
+        if not voice_client:
+            return
+
+        def after_interrupt(_error):
+            if current_source:
+                voice_client.play(current_source, after=self.after_callback)
+            self.interrupt_play_lock.release()
+
+        current_source = voice_client.source
+        if not self.interrupt_play_lock.locked():
```

Comment:
Using a lock here can put the bot into an unrecoverable state. If the after callback is for some reason not called this lock is never released meaning we will never be able to interrupt_play again until we restart the bot.

To fix this you could maybe instead have a lock for entering interrupt_play, this way interrupt_play would just stack successive audio calls i.e. ringing the dinkster when ringing the dinkster will stop the first dinkster and then resume it later and then finally return to the original audio.

Reply:
I tried this, but dinkstering a running dinkster cancels resuming playback, and resumes only the last dinkster interrupted :(

Reply:
I guess the reason for this is that 'after_interrupt' passes self.after_callback as the after callback, instead of the after callback of the first dinkster (line 471). Could be fixed with a stack or storing the current after callback in a member variable.

Reply:
Removed all locks. Multiple interrupts is unhandled. Multiple interrupts is undefined behaviour :stuck_out_tongue_winking_eye:

Reply:
Nice x)